### ISSUE TITLE
Consider changing the precedence of title sources

### ISSUE THREAD

**Issue by ofek** [mkdocs/mkdocs#3532]
Specifically, on this page https://www.mkdocs.org/user-guide/writing-your-docs/#meta-data I would prefer that the level 1 headers take priority over the navigation name, or an option to do so.

I'm realizing that deriving the title from the navigation sidebar does not provide a good user experience and I would like to avoid having to add meta-content to every page. I will use a real example: https://hatch.pypa.io/1.9/config/static-analysis/

The sidebar already represents clearly defined sections and in this case the entire sidebar relates to config so it would be very verbose to add the word "configuration" to every name.

Here is the full(ish) picture showing the desired title as the level 1 header:

![image](https://github.com/mkdocs/mkdocs/assets/9677399/119435e1-1ed1-425b-910c-e864ab48f176)

A consequence is that when you scroll down with Material for MkDocs the title on top is now no longer specific:

![image](https://github.com/mkdocs/mkdocs/assets/9677399/f2f3bccc-e67d-408c-97e1-951b63188c19)

As well as the social card that shows up when you post a link to a page in Slack or Twitter or Discord:

![slack-imgs](https://github.com/mkdocs/mkdocs/assets/9677399/572e41bd-bb7a-4deb-b46d-73e90add1f7b)

**Comment by ofek** [comment]
Maybe there is a way to achieve this immediately with a hook?

**Comment by oprypin** [comment]
I think this is a behavior of mkdocs-material, or I don't understand what you're describing.

**Comment by oprypin** [comment]
Aha no, I see. Although both of the shown examples of how the page title is displayed are unique to mkdocs-material, the source of that title is deduced by MkDocs. It's presumably also put into the `<title>` tag which would be a generic example?

**Comment by oprypin** [comment]
Yeah it's a good point. MkDocs just expects you to use the same title in the navigation and in the first heading. A page doesn't have multiple kinds of titles. This feature would be tough to introduce conceptually and would have to be an awkward new config entry 😞

**Comment by oprypin** [comment]
Actually all themes at the moment explicitly use the same `title` attribute for both the page title and the nav entry. So in a simplistic view, this isn't even possible to change by changing something only in MkDocs.

https://github.com/mkdocs/mkdocs/blob/7186f4ce9c88e7c445ef90fee85c065ca297a7c8/mkdocs/themes/mkdocs/base.html#L16

https://github.com/mkdocs/mkdocs/blob/7186f4ce9c88e7c445ef90fee85c065ca297a7c8/mkdocs/themes/mkdocs/base.html#L100

My only idea is to add a new attribute and then themes could use it. There actually already is such an attribute but it's private - 
https://github.com/mkdocs/mkdocs/blob/7186f4ce9c88e7c445ef90fee85c065ca297a7c8/mkdocs/structure/pages.py#L281

**Comment by ofek** [comment]
cc @squidfunk for visibility

**Comment by ofek** [comment]
> Actually all themes at the moment explicitly use the same `title` attribute for both the page title and the nav entry.

In Material for MkDocs if I set the following then the title is in fact updated appropriately without changing the text in the sidebar: https://squidfunk.github.io/mkdocs-material/reference/#setting-the-page-title

Is there some way I can artificially set this metadata attribute in a hook based on the level 1 header?

**Comment by squidfunk** [comment]
Title handling is broken since MkDocs 1.5 (see #3357). There are some instances where markup ends up in `Page.title` (see linked issue). Should `Page.meta.title` be allowed to contain markup? Plugins that rely on reading page titles in `on_page_markdown` or `on_page_content` might end up with inconsistent results, if `Page.title` is changed after reading the Markdown, but I guess this is no more worse than the current situation. Before anything is changed on title handling, it should be solidly spec'd, because there are several cases that currently lead to undefined behavior.

My 2 cents.

**Comment by ofek** [comment]
https://github.com/pypa/hatch/pull/1216

I just tried setting the front matter via a hook but it looks like at that point the metadata is already stripped out and processed. Is there a way to set the title based on the h1 via a hook?

**Comment by ofek** [comment]
@oprypin Do you know of a way?

**Comment by oprypin** [comment]
@ofek It is **extremely** convoluted to bypass the title machinery right now in the way that you want, which is why I wasn't suggesting anything. Could you wait for further developments regarding titles

**Comment by ofek** [comment]
Sure I'm fine with waiting sorry to bother! Although, outside of titles is there any way to change front matter via a hook?

**Comment by oprypin** [comment]
> Should `Page.meta.title` be allowed to contain markup

No but it always was allowed to contain markup :disappointed: and seems like it will stay this way. You've been saying it like MkDocs 1.5 changed this but actually it didn't change anything for `Page.meta.title`. Indeed, see https://github.com/mkdocs/mkdocs/issues/3357#issuecomment-1925438350

**Comment by oprypin** [comment]
For this change itself, I am interested to know if you'd be happy to apply it when the new attribute is available, @squidfunk .

* https://github.com/mkdocs/mkdocs/pull/3533

https://github.com/squidfunk/mkdocs-material/blob/bcad9cecd18b47409988ebb0aa69bea6a7edd70c/src/templates/base.html#L91-L97
to become
```jinja
      {% if page.title and not page.is_homepage %}
        <title>{{ page.content_title | striptags }} - {{ config.site_name }}</title>
      {% else %}
        <title>{{ config.site_name }}</title>
      {% endif %}
```

https://github.com/squidfunk/mkdocs-material/blob/bcad9cecd18b47409988ebb0aa69bea6a7edd70c/material/templates/partials/header.html#L28-L32
to become
```jinja
            {{ page.content_title }}
```

You've added this special case to prioritize as follows:
1. meta['title']
2. title from mkdocs.yml `nav`
3. title from content

But instead this would now prioritize as follows:
1. meta['title']
2. title from content
3. title from mkdocs.yml `nav`

Which would make more sense to both me and @ofek

**Comment by squidfunk** [comment]
Honestly, I'm having a really hard time committing to anything with "yes" or "no" right now, because there are so many changes related to title handling in the making that I'm not able to keep up. What you're proposing _sounds_ good, but that is only a guess. There might be side effects that break things. If this gets merged, IMHO, the maintainers of MkDocs might consider a prerelease to give plugin an theme authors enough time to test and adjust.

**Comment by oprypin** [comment]
Thanks for the feedback. I still believe that there's no way that the current behavior makes more sense than the proposed behavior, but with this feedback in mind, it will surely break at least some users' expectation from pre-existing behavior. Then perhaps my change should stop at just adding the new attribute, without changing built-in themes.

**Comment by oprypin** [comment]
And I do think that at the very least what mkdocs-material currently does would make a lot of sense for all themes. Namely,
```jinja
<title>{{ page.meta.title or page.title }}</title>
```
because with similar logic, why on earth would the user specify the meta title if it's impossible for it to be used if the site specifies titles in the nav config. 
I think this should be recommended and built-in themes should be changed to this

**Comment by ofek** [comment]
> I suppose an additional config option could be added which basically causes nav titles to be ignored. But then why is the user defining them?

Please see the pictures I posted in the OP. It is desirable to have the actual title be more verbose which is undesirable in the short link one would click on the side.

**Comment by oprypin** [comment]
Simply adding the new attribute will enable mkdocs-material to change the source of this floating pop-out title element. (Which is a totally unique concept in this theme.) I think that will be uncontroversial.

 The question whether it will also make sense change the source of the `<title>` is not so simple apparently.

**Comment by pawamoy** [comment]
OK, I had to re-read a few times to make sure I understand. If the proposal is the following, then I'm -0:

- titles specified in mkdocs.yml's `nav` keep being used as labels for the navigation (left sidebar in Material for MkDocs)
- HTML `<title>` for each page is now populated from the H1 heading of the page, not the title specified in mkdocs.yml's `nav`
- in cases where this is not desirable, it is still possible to override the HTML `<title>` with the front-matter `title`metadata

In some pages of mine, I have the opposite case of @ofek's: Markdown title is short, nav title is longer and more descriptive, so I want the nav title to be used as HTML title (current behavior). With this proposal implemented, I can still achieve that by copying the nav title in the front-matter `title` metadata, so that's not all bad. IMO at this point, it doesn't really matter which of the Markdown or nav title is used for the HTML title: some users will want the first, others the second (I myself use both :shrug:). I'm not sure which one makes the most sense. Since not all users will be satisfied, I think I'd favor not breaking current behavior? Maybe some usage statistics of front-matter `title` metadata would give us some pointers?

**Comment by pawamoy** [comment]
(I'm also therefore totally open to a new option for giving precedence to nav or Markdown titles)

**Comment by oprypin** [comment]
Alright, so there will be no change to how `<title>` is determined by default.
https://github.com/mkdocs/mkdocs/pull/3533 will only keep the addition of the new attribute but not use or recommend it for anything.

This can enable more flexibility in hooks (example https://github.com/pypa/hatch/pull/1239) and new possibilities for themes with advanced features (example https://github.com/mkdocs/mkdocs/issues/3532#issuecomment-1925988186)

**Comment by oprypin** [comment]
Or well actually I am still interested to bump the priority of page.meta.title over the nav title , like mkdocs-material already does
 (https://github.com/mkdocs/mkdocs/issues/3532#issuecomment-1925973278)
But definitely not bumping the priority of the H1 title

**Comment by waylan** [comment]
> why on earth would the user specify the meta title if it's impossible for it to be used if the site specifies titles in the nav config

Some more background. MkDocs didn't always to support defining page titles in the meta-data. And so if a user wanted a page title to be something other than the the `<h1>`, then they needed to use the nav. However, understandably, users did not like editing page attributes somewhere other than in the page file itself. So we added the meta-data option for those users.

So nav was the oldest and only option to begin with. (with filename as a fallback) Those users need to continue to be supported. Each additional option was built later, so they all have lower priority. And as explicit takes precedence over implicit, then we get `nav, meta-data, content, filename` as the only order we can support without breaking existing users sites.

But yes, a theme (or a custom template provided by a user) could conceivably check `page.meta.title` and use that instead of `page.title`. However, that only makes sense when `page.title` comes from the nav. Because in any other scenario, `page.meta.title` and `page.title` will be the same. That said, if a theme used some different meta-data key (like maybe `nav-title`), then anything is possible (including `<title>{{ page.meta.nav-title or page.title }}</title>`. In fact, this doesn't even need any changes to MkDocs for the user to do. A theme could build this in or even use a theme config option to optionally enable it. Or a user could define their own [custom template block](https://www.mkdocs.org/user-guide/customizing-your-theme/#overriding-template-blocks) to override the relevant block of the theme they use.

**Comment by ofek** [comment]
> In some pages of mine, I have the opposite case of @ofek's: Markdown title is short, nav title is longer and more descriptive, so I want the nav title to be used as HTML title (current behavior).

@pawamoy Can you please share an example? I think in the vast majority of cases users would benefit from the default sourced from the `h1`. I understand that probably won't happen, I'm just saying.

**Comment by pawamoy** [comment]
And you might be right! It would be interesting to look at current usage of meta+nav titles (only working with mkdocs-material IIUC) compared to nav+markdown titles, to see which use is the most common. I'm going to bed but I'll share an example tomorrow :)

**Comment by oprypin** [comment]
> Alright, so there will be no change to how `<title>` is determined by default.
> https://github.com/mkdocs/mkdocs/pull/3533 will only keep the addition of the new attribute but not use or recommend it for anything.
> 
> This can enable more flexibility in hooks (example https://github.com/pypa/hatch/pull/1239) and new possibilities for themes with advanced features (example https://github.com/mkdocs/mkdocs/issues/3532#issuecomment-1925988186)

* I only now actually updated the code in https://github.com/mkdocs/mkdocs/pull/3533 accordingly

**Comment by squidfunk** [comment]
> That's something that I think? goes way back to early versions of MkDocs. A much neater approach would be to just be explicit. A nav config that always includes titles would be how I'd approach this if I was looking at a redesign. The toc structure could also be passed as context, and the decision around which to us would be a theming issue.

I just had a talk with @pawamoy on another backlog grooming call that we did today on this very topic, after we didn't have enough time yesterday. It's important to know that there are many users that just throw a folder of Markdown files at MkDocs without specifying an explicit structure, which is an awesome way to get started quickly.

There are also several plugins that allow to manage explicit navigation auto-magically outside of `mkdocs.yml`. Thus, keeping this functionality, IMHO, is absolutely essential to provide a great out-of-the box experience ☺️ Also note that it's not primarily a theming issue, as plugins also need to know titles of pages, e.g. the search plugin.

All of that just from the top of my head – deeper evaluation required.

**Comment by tomchristie** [comment]
> It's important to know that there are many users that just throw a folder of Markdown files at MkDocs without specifying an explicit structure

Related #3356

**Comment by honzajavorek** [comment]
It seems to me that with the upgrade to 1.6.0 the `page.title` attribute now gives me whatever I've written to the front matter as a `title:`, is that correct? Because that completely breaks how I've been using MkDocs until now. I need three distinct titles:

- **nav title**, a short one which displays in the navigation, footer of the website, etc.
- **SEO title**, the one which displays in the `<title>` tag and thus in Google
- **H1 title**, the one which displays as the largest one to the actual reader of the page

I'm using a custom theme. Until 1.6.0 I believe the mechanics of titles worked the way that in theme, `page.title` gave me the nav title and `page.meta.title` gave me the SEO title. I didn't need the H1 title in the theme for anything, thus no regrets it wasn't accessible.

Now it seems that `page.title` and `page.meta.title` return the same thing, which breaks everything I have. It wouldn't be an issue if I could use a different attribute instead to access the nav title, but I see no such thing. I'm reading code of `Page` and scanning through this issue, but it seems the only way to fix this for me is to pin to an old version of MkDocs and then figure out what to do and if writing a custom plugin or something would help me.

Was such change of behavior intended or is it a bug?

**Comment by honzajavorek** [comment]
I can now confirm that downgrading to 1.5.3 solves the issue. This change of behavior isn't mentioned in the 1.6.0 changelog (only that HTML gets parsed etc.), so it seems like a bug to me, but YMMV. If you want me to go to a separate issue with this, let me know and I'll do that and delete these comments not to pollute this one.

**Comment by pawamoy** [comment]
Yes, please open a new issue with a minimal reproducible example so that it is easier for us to investigate. I did a quick test and `page.title` gives me the title defined in `nav`, not in the front-matter. `page.meta` is empty though :thinking:

**Comment by squidfunk** [comment]
I think https://github.com/squidfunk/mkdocs-material/issues/7411 might also be related.

**Comment by honzajavorek** [comment]
Today I rolled up my sleeves to debug this. I made a minimal example and everything. And I couldn't reproduce the issue I previously described. So I added `mkdocs-awesome-pages-plugin` as that's what I use on production. I still couldn't reproduce the issue. So I tried to upgrade `mkdocs` at the production repo to see what actually happens, and I couldn't reproduce the issue. Titles work for me now 🤷‍♂️

A while back I changed the code which generates my navigation, for different reasons. I suspect that perhaps the code worked in a way which was all right with MkDocs 1.5, but wouldn't work with MkDocs 1.6? And when I changed it, I accidentally fixed the issue? Who knows… But for now, I guess I'm out!
